Implement interfaces for RowExpressionTranslation

sachdevs opened the pull request:
Implement Row Expression translation interfaces in new module presto-expression.

NOTE: See an example implementation using these interfaces in #13526 

```
== NO RELEASE NOTE ==
```

highker commented on presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/optimization/JdbcComputePushdown.java:
```diff
@@ -105,7 +112,12 @@ public PlanNode visitFilter(FilterNode node, Void context)
+                    rowExpressionTranslator,
+                    node.getPredicate(),
+                    new DefaultRowExpressionTranslator.ConnectorContext(oldTableScanNode.getAssignments()));
+            // TODO: put the translated filter into JdbcTableLayoutHandle
```
Let's implement this and add the corresponding logic in `com.facebook.presto.plugin.jdbc.QueryBuilder`.

highker commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/FunctionTranslationFromAnnotationsParser.java:
```diff
@@ -0,0 +1,200 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class FunctionTranslationFromAnnotationsParser
```
Can we explore the possibility of converging this with `ScalarFromAnnotationsParser`. For example `ScalarHeaderAndMethods` is common. Also, `ScalarTranslationFromAnnotationsParser` could be a better name

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ConnectorPlanOptimizerManager.java:
```diff
@@ -16,35 +16,88 @@
+    private final FunctionMetadataManager functionMetadataManager;
+    private final Map<ConnectorId, ConnectorPlanOptimizerProvider<?>> planOptimizerProviders = new ConcurrentHashMap<>();
+
+    // TODO: make this lazy like planOptimizerProviders
```
Check the possibility of loading this lazily like `planOptimizerProviders`.

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorPlanOptimizerProvider.java:
```diff
@@ -14,10 +14,67 @@
+     * @param functionTranslator that maps CallExpression to TargetExpression.
+     * @return a set of optimizer rules.
+     */
+    Set<ConnectorPlanOptimizer> getConnectorPlanOptimizers(TranslationContext<T> functionTranslator);
```
Maybe we should just use `TranslationContext<T>` to avoid templated `ConnectorPlanOptimizerProvider<T>`. Check how other connector interfaces solve the problem. @hellium01 can help on this front.

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/relation/DefaultRowExpressionTranslator.java:
```diff
@@ -0,0 +1,75 @@
+
+    public static class ConnectorContext
+    {
+        Map<VariableReferenceExpression, ColumnHandle> columnHandleMap;
```
I wouldn't put this in spi. It is only used by translateVariableReference. We can remove this and ask connector to inherit DefaultRowExpressionTranslator

hellium01 commented:
@sachdevs , let's sync up offline? I made some changes on top @highker and your code, which I hope can make discussing it a little easier?
https://github.com/hellium01/presto/commit/fa01300325f58d445f05a7c335d32e6c581608da

sachdevs commented:
Added Yi's changes and now addressing James' comments + cleaning up further.

sachdevs commented:
Added tests, ensured basic code paths work correctly, and cleaned up further.
Functionality TODOs:
- Add optimizer flag to Jdbc (for turning on/off feature)
- Add JdbcOptimizerProvider to executed code path so that the optimizer can be used if desired
- Add other Jdbc translation functions? @hellium01 how far do we want to go for the JDBC example?

- Split this into multiple commits
- Add integration tests for end-to-end verification
- Add more javadocs for SPI, clean up further

sachdevs commented:
Added another test suite, now splitting up commits + cleaning up.

sachdevs commented:
Feel free to take a look for any high-level mistakes/ things I'm still missing before I start to get deeper into implementing the JDBC connector using this interface.

rongrong commented:
How confident are we about this is the right approach to do translation and these are the right interfaces? Should we consider putting these in a separate package rather than add all them to spi until we get better understanding and confidence that these are the right things to do?

sachdevs commented:
After discussing with James, moving to a separate module and decoupling SPI from this PR.

sachdevs commented:
Refactored to be in separate new top-level library module `presto-expressions`. Moved Jdbc implementation commit to separate PR. Let me know if I was correct in assuming ConnectorPlanOptimizerManager goes into the new module as well.

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/ConnectorPlanOptimizerManager.java:
```diff
@@ -0,0 +1,50 @@
+import static com.google.common.collect.Maps.transformValues;
+import static java.util.Objects.requireNonNull;
+
+public class ConnectorPlanOptimizerManager
```
We don't need this class in the new package

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/annotations/ScalarImplementationHeader.java:
```diff
@@ -0,0 +1,117 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ScalarImplementationHeader
```
- Call this `ScalarTranslationHeader`
- Make it a private static inner class of `TranslatorAnnotationParser`

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/annotations/TranslatorAnnotationParser.java:
```diff
@@ -0,0 +1,200 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.expressions.annotations;
```
Move this to `translator` dir

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/FunctionTranslator.java:
```diff
@@ -0,0 +1,74 @@
+    {
+        ImmutableMap.Builder<FunctionMetadata, MethodHandle> functionMappingBuilder = new ImmutableMap.Builder<>();
+        translatorContainers
+                .stream()
```
Put `.stream()` in the previous line

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/annotations/TranslatorAnnotationParser.java:
```diff
@@ -0,0 +1,200 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class TranslatorAnnotationParser
```
Make it package-private once moved to `translator` dir

sachdevs commented:
Cleaned up, adding unit tests now.

sachdevs commented:
Finished writing tests. Let me know if we would like more coverage on any specific code path. Currently I'm hitting most parts of the translator and annotation parser.

highker commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,224 @@
+        this.sqlToRowExpressionTranslator = new TestingRowExpressionTranslator(METADATA);
+    }
+
+    private class TestNonRecursiveExpressionTranslator extends RowExpressionTranslator<String, Map<VariableReferenceExpression, ColumnHandle>> {
```
Nit: we always put private class at the end of a file; also break a line:
```java
private static class TestNonRecursiveExpressionTranslator
            extends RowExpressionTranslator<String, Map<VariableReferenceExpression, ColumnHandle>>

highker commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    @Test
+    public void testNonRecursiveTranslator() {
```
put { in it's own line

highker commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+
+    private class TestFunctionTranslator extends RowExpressionTranslator<String, Map<VariableReferenceExpression, ColumnHandle>> {
```
same nit

highker commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    private static final FunctionHandle TEST_FUNCTION = new TestFunctionHandle();
```
nit: no need to create the test class; use lambda:
```java
private static final FunctionHandle TEST_FUNCTION = () -> FullyQualifiedName.of("a.b.c").getPrefix();

sachdevs replied:
Cool forgot you could do that :)

highker commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    @Test
+    public void testEndToEndFunctionTranslation2() {
```
Combine `testEndToEndFunctionTranslation` and `testEndToEndFunctionTranslation2` into one

highker commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,224 @@
+    private static class TestFunctions {
+        @ScalarFunction
+        @SqlType(StandardTypes.BIGINT)
+        public static String bitwiseAnd(@SqlType(StandardTypes.BIGINT) TranslatedExpression<String> sql1, @SqlType(StandardTypes.BIGINT) TranslatedExpression<String> sql2)
```
nit: sql1 -> left; sql2 -> right

highker commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,224 @@
+        @SqlType(StandardTypes.BIGINT)
+        public static String bitwiseAnd(@SqlType(StandardTypes.BIGINT) TranslatedExpression<String> sql1, @SqlType(StandardTypes.BIGINT) TranslatedExpression<String> sql2)
+        {
+            return sql1.getTranslated().get() + " BITWISE_AND " + sql2.getTranslated().get();
```
assert getTranslated is present

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/FunctionTranslator.java:
```diff
@@ -0,0 +1,72 @@
+        this.functionMapping = functionMapping;
+    }
+
+    public boolean isFunctionPresent(FunctionMetadata functionMetadata)
```
private

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/FunctionTranslator.java:
```diff
@@ -0,0 +1,72 @@
+        return new FunctionTranslator<>(functionMappingBuilder.build());
+    }
+
+    private FunctionTranslator(Map<FunctionMetadata, MethodHandle> functionMapping)
```
move private functions after public ones

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/FunctionTranslator.java:
```diff
@@ -0,0 +1,72 @@
+        return functionMapping.containsKey(functionMetadata);
+    }
+
+    public Optional<T> translate(FunctionMetadata functionMetadata, List<TranslatedExpression<T>> translatedArguments)
```
I think this should return `TranslatedExpression<T>`. Could you add test to cover this?

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTranslator.java:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * translator for function that is recursive. Will be overidden if non-recursive version returns translated values.
+     */
+    public Optional<T> translateCall(FunctionHandle functionHandle, List<TranslatedExpression<T>> translatedExpressions, C context)
```
All these calls should return `TranslatedExpression<T>`. It is up to TranslatedExpression to decide if they are optional empty or not.
Also, why we have two translateCalls and translateSpecialForms? We don't really need the ones handling `functionHandle` or  `SpecialFormExpression.Form`

sachdevs replied:
Yeah I agree. That makes sense, refactored as such.

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTreeTranslator.java:
```diff
@@ -0,0 +1,121 @@
+        public TranslatedExpression<T> visitCall(CallExpression call, C context)
+        {
+            Optional<TranslatedExpression<T>> defaultTranslated = rowExpressionTranslator.translateCall(call, context, RowExpressionTreeTranslator.this);
+            if (defaultTranslated.isPresent()) {
```
nit: use `return defaultTranslated.orElse(...);`

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTranslator.java:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * default translator for expression that is non-recursive.
+     */
+    public Optional<TranslatedExpression<T>> translateExpression(RowExpression expression, C context, RowExpressionTreeTranslator<T, C> rowExpressionTreeTranslator)
```
this should be private:
```java
private TranslatedExpression<T> translateExpression(RowExpression expression, C context, RowExpressionTreeTranslator<T, C> rowExpressionTreeTranslator)
{
    return TranslatedExpression.untranslated(expression);
}

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTranslator.java:
```diff
@@ -0,0 +1,77 @@
+
+public class RowExpressionTranslator<T, C>
+{
+    /**
```
We don't really need these javadoc; the interfaces are pretty self-explanatory.

sachdevs commented:
Looks like I missed the checkstyle run, on my last revision, fixed those issues now.

The refactor also resulted in me getting rid of the tests that check the "automatic" subtree translation by the library. Let me know if there are any other tests I am missing. Is it worthwhile to unit test annotation parsing in isolation? I.e. make a temp class to parse and ensure correct entries show up in the functionMapping?

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTranslator.java:
```diff
@@ -0,0 +1,50 @@
+        return translateExpression(specialForm, context, rowExpressionTreeTranslator);
+    }
+
+    private TranslatedExpression<T> translateExpression(RowExpression expression, C context, RowExpressionTreeTranslator<T, C> rowExpressionTreeTranslator)
```
Inline this function to `translateCall ` and `translateSpecialForm `

sachdevs replied:
Yeah, removing since translateExpression since we no longer do recursive translation.

As for merging RowExpressionTranslator and RowExpressionTree translator, I see the argument for why that may make sense since we do not do that much extra logic in tree translator anymore. However, in terms of an easy-to-understand interface for the developer I find that the way it is probably would work better since at a glance the developer knows what they need to extend/implement. Otherwise we would essentially be making them write their own `RowExpressionVisitor` if we merged these classes.

highker replied:
Just remove it. Not that useful outside this class.

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTranslator.java:
```diff
@@ -0,0 +1,50 @@
+
+import static com.facebook.presto.expressions.translator.TranslatedExpression.untranslated;
+
+public class RowExpressionTranslator<T, C>
```
Let's put translateLambda as well

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTreeTranslator.java:
```diff
@@ -0,0 +1,96 @@
+        @Override
+        public TranslatedExpression<T> visitInputReference(InputReferenceExpression reference, C context)
+        {
+            throw new UnsupportedOperationException("Cannot translate RowExpression that contains inputReferenceExpression");
```
Add a comment: "InputReferenceExpression should only be used by Presto engine rather than connectors"

highker commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTreeTranslator.java:
```diff
@@ -0,0 +1,96 @@
+        @Override
+        public TranslatedExpression<T> visitLambda(LambdaDefinitionExpression lambda, C context)
+        {
+            return TranslatedExpression.untranslated(lambda);
```
Call `rowExpressionTranslator.translateLambda`

hellium01 commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTreeTranslator.java:
```diff
@@ -0,0 +1,96 @@
+            RowExpressionTranslator<T, C> translator,
+            C context)
+    {
+        LogicalRowExpressions logicalRowExpressions = new LogicalRowExpressions(determinismEvaluator, functionResolution, functionMetadataManager);
```
We don't need to put this logic in the `RowExpressionTreeTranslator`, the user of `translateWith` need to do before they call `translateWith`.

sachdevs replied:
Are there any side-effects of this assumption? I.e. if the user does not convert the expression/predicate to conjunctive form, could their be any resultant issues other than building the *minimal* converted rowexpression? I just want to make sure that by moving this out we dont make the code less fool-proof.

hellium01 replied:
I don't think this `convertToConjunctiveNormalForm` is needed for all translations, it is only used to make predicate pushdown easier, so definitely remove it from the common interface.

hellium01 commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/FunctionTranslator.java:
```diff
@@ -0,0 +1,75 @@
+    public TranslatedExpression<T> translate(FunctionMetadata functionMetadata, RowExpression original, List<TranslatedExpression<T>> translatedArguments)
+    {
+        if (!isFunctionPresent(functionMetadata)) {
+            return untranslated(original);
```
return:
```
   return new TranslatedExpression<>(Optional.empty(), original, translatedArguments);
```
?  the information of translatedArguments might be lost here.

sachdevs replied:
Nice catch, i guess we have no reason to throw away translated arguments

hellium01 commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/FunctionTranslator.java:
```diff
@@ -0,0 +1,75 @@
+            return new TranslatedExpression<>(Optional.of((T) functionMapping.get(functionMetadata).invokeWithArguments(translatedArguments)), original, translatedArguments);
+        }
+        catch (Throwable t) {
+            return untranslated(original);
```
I agree this should be an error.

hellium01 commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/FunctionTranslator.java:
```diff
@@ -0,0 +1,75 @@
+
+    private boolean isFunctionPresent(FunctionMetadata functionMetadata)
+    {
+        return functionMapping.containsKey(functionMetadata);
```
We probably can inline this function.

hellium01 commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Test
+    public void testEndToEndFunctionTranslation()
```
Can you also cover:
1) Function that does not have a rule, so we cannot translate?
2) Function has same name but different signature (for example input types are different)?
3) Function that is hidden, probably `not`.
4) Simple operators such as: +, -, > .

hellium01 commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Test
+    public void testEndToEndSpecialFormTranslation()
```
Can you also cover special form that cannot be translated?

hellium01 commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,182 @@
+    @Test
+    public void testEndToEndFunctionTranslation()
+    {
+        String untranslated = "LN(bitwise_and(1, col1))";
```
We also might want to cover:
some types of constant that is not translatable.

sachdevs replied:
what type of constant would that be? untranslatable constants wouldnt make it past expression creation no? since the type provider entry will be missing?

hellium01 commented on presto-expressions/src/main/java/com/facebook/presto/expressions/translator/TranslatorAnnotationParser.java:
```diff
@@ -0,0 +1,292 @@
+    {
+    }
+
+    public static Map<FunctionMetadata, MethodHandle> parseFunctionDefinitions(Class<?> clazz)
```
I would prefer we do some checking on return type and argument type during `parseFunctionDefinition` rather than throw later when we call translation if we have wrongly defined translate functions.

sachdevs replied:
Yeah I can see why that is useful. Makes the code more defensible and we fail earlier.

sachdevs commented:
Updated PR with comments from @highker and @hellium01. Just looking for verification on what we need to do about the unresolved comments and lets finish this up 😄

sachdevs commented:
Done, removed those params.

highker commented on presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java:
```diff
@@ -0,0 +1,300 @@
+        public TranslatedExpression<String> translateCall(CallExpression callExpression, Map<VariableReferenceExpression, ColumnHandle> context, RowExpressionTreeTranslator<String, Map<VariableReferenceExpression, ColumnHandle>> rowExpressionTreeTranslator)
+        {
+            List<TranslatedExpression<String>> translatedExpressions = callExpression.getArguments().stream()
+                    .map(expression -> expression.accept(rowExpressionTreeTranslator.getVisitor(), context))
```
use `rewrite` than `getVisitor`
